### PR TITLE
feat(contract-distribution): Exclude old-code from witness for deploy-contract and delete-account actions

### PR DESCRIPTION
**Pull request by tayfunelmas**
Today we record contract accesses for function calls, so we do not include the contract code in the state witness, instead we distribute these contracts separately to the validators.

There are also two other actions (`DeployContract` and `DeleteAccount`) that are seen less frequently but also reads code from the storage and contribute to state witness.

These actions do not run the code read from storage but only capture the code length to update storage usage. Thus, there is no need to add the actual code to the validators in the witness. To address this, we change the implementation of these actions to read the code length without reading the actual code.

Related changes in this PR:

- As a refactor, move `get_code` and `set_code` function from `core\store\src\lib.rs` to `TrieUpdate` as member methods. Note that, this will potentially be leveraged to change the code further to use `ContractStorage` instead of the trie and `prospective` changes.

- Add a new method to `TrieUpdate` called `get_code_len` to call `get_ref` to access the reference and capture the length of the value. Update the `DeployContract` and `DeleteAccount` actions to use this new method instead of reading the whole code.

- Small refactor: Change visibility of `contract_storage`.

- Add new tests for deleting the account after deploying and calling contracts on them. For this, we use accounts separate from the validator accounts.

### CODE REVIEW

File: core/store/src/trie/update.rs [hunk 1]
```diff
@@ -139,6 +155,44 @@ impl TrieUpdate {
...
+        account_id: AccountId,
+        code_hash: CryptoHash,
+    ) -> Result<Option<usize>, StorageError> {
+        let key = TrieKey::ContractCode { account_id };
+        let value_ptr = self.get_ref(&key, KeyLookupMode::FlatStorage)?;
+        if let Some(value_ptr) = value_ptr {
+            let value_hash = value_ptr.value_hash();
+            debug_assert_eq!(
+                code_hash, value_hash,
+                "Hash for code does not match hash in account: Expected: {:?} Found: {:?}",
```
Comment (pugachAG) [inline core/store/src/trie/update.rs:182]: nit: assert_eq should include expected and actual values in the error message, no need to manually includ eit

File: runtime/runtime/src/actions.rs [hunk 2]
```diff
@@ -618,9 +617,10 @@ pub(crate) fn action_deploy_contract(
 ) -> Result<(), StorageError> {
     let _span = tracing::debug_span!(target: "runtime", "action_deploy_contract").entered();
     let code = ContractCode::new(deploy_contract.code.clone(), None);
```
Comment (pugachAG) [inline runtime/runtime/src/actions.rs:619]: missing protocol version check

File: runtime/runtime/src/actions.rs [hunk 3]
```diff
@@ -660,10 +660,10 @@ pub(crate) fn action_delete_account(
     if current_protocol_version >= ProtocolFeature::DeleteActionRestriction.protocol_version() {
         let account = account.as_ref().unwrap();
         let mut account_storage_usage = account.storage_usage();
-        let contract_code = get_code(state_update, account_id, Some(account.code_hash()))?;
```
Comment (pugachAG) [inline runtime/runtime/src/actions.rs:663]: missing protocol version check here as well

**Review by pugachAG — CHANGES_REQUESTED** [review]
LGTM, just need to add protocol version checks to preserve backward compatibility

File: core/store/src/trie/update.rs (hunk 1 above)
Comment (tayfunelmas) [reply core/store/src/trie/update.rs]: done.

File: core/store/src/trie/update.rs [hunk 4]
```diff
@@ -291,21 +349,16 @@ impl TrieUpdate {
...
             })?;
-        let contract_exists: bool = match contract_ref {
-            Some(OptimizedValueRef::Ref(value_ref)) => value_ref.hash == code_hash,
-            Some(OptimizedValueRef::AvailableValue(ValueAccessToken { value })) => {
-                hash(value.as_slice()) == code_hash
-            }
-            None => false,
-        };
+        let contract_exists =
+            contract_ref.map(|value_ref| value_ref.value_hash() == code_hash).unwrap_or(false);
```
[the remark this answers is not in the document]
Comment (tayfunelmas) [reply core/store/src/trie/update.rs:352]: thanks, done.

File: runtime/runtime/src/actions.rs (hunk 2 above)
Comment (tayfunelmas) [reply runtime/runtime/src/actions.rs]: done.

File: runtime/runtime/src/actions.rs (hunk 3 above)
Comment (tayfunelmas) [reply runtime/runtime/src/actions.rs:663]: done

**Comment by tayfunelmas** [comment]
Added protocol version checks.

**Review by pugachAG — APPROVED** [review]

**Comment by tayfunelmas** [comment]
> If we record just the length, does that ensure that the state is consistent globally and whoever produced the witness isn't being dishonest? IOW do we still record trie nodes leading up to the contract value so that the contract code size contributes to the state root?

Before this change, we record the following in the witness:

1. The trie nodes from state-root up to (and including) the leaf node pointing to the value (code). This leaf node contains the size and hash of the code, thus they both contribute to the state root. (https://github.com/near/nearcore/blob/f96373207b3faa90bbfa89f689f3eefc43f5c9d2/core/primitives/src/state.rs#L8-L12)
2. The value (code) itself pointed by the leaf node.

With this change, we still traverse the trie up to the leaf and send (1) but we do not deref the value so we do not sent (2).

If this was for an action that sets the size in the post-state, not sending the code would be a problem. However, note that, in both deploy and delete-account actions, the code that we skip to send is already in the pre-state of the state transition (code that was deployed to that account before), so, the size and hash of the code should be proved in previous state transitions leading to that state root. Or, if the deploy of the existing code is also in the same chunk, then we should already be distributing that code to the validator in the receipts in the witness.

In other words, if the code-size in the leaf is not correct, the state should already been corrupted.

Let me know if this makes sense, or I miss something.

@pugachAG to see if I miss anything as well.
